### PR TITLE
Create amount filter

### PR DESCRIPTION
**Pull request by SzymczakJ**
### Details

This PR adds an amount filter.

### Fixed Issues

$ https://github.com/Expensify/App/issues/46036
PROPOSAL:

### Tests

1. Go to advanced filters tab
2. Try setting up amount filter and make sure the inputs work correctly 
Screenshot 2024-08-14 at 10 47 29
Screenshot 2024-08-14 at 10 48 53

3. Save and view results and verify that values are filtered by amount.
Screenshot 2024-08-14 at 10 55 45
Screenshot 2024-08-14 at 10 56 23

### QA Steps

Same as tests

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/30def578-47e7-499a-bcd3-7ec78234f41f

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/dc9947eb-f0bf-4e32-a61b-8b2dd7ad7b97

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/d7f849a6-ab81-4155-8d85-dbb6b192336c

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/bec8126a-f686-4677-8117-e18dea3ad331

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/311d0ced-fe21-41e2-8c91-3b0c91961ccb

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/60abebc7-4ffc-4108-87b9-c4266f81ebad

</details>

### CODE REVIEW

**Comment by SzymczakJ** [comment]
@luacmartins I noticed that when we set url with param `amount>2000` the BE treats 2000 as if this value was 20.00.
I know that this approach was is widely used on BE, but I think that in this case we should treat 2000 as 2000.00. 
We don't specify currency and there are some currencies with no digits after decimal point(like JPY), so when somebody would want to filter out 20 JPY he would get values under 0.20 JPY and this doesn't make sense.
Can we change it on BE side 🙏 ?

**Review by 289Adam289 — APPROVED** [review]
One small problem, but other then that LGTM!

**Comment by SzymczakJ** [comment]
I'll OOO the whole next week but I asked @289Adam289 to take care of merging this one 😄

File: src/components/AmountWithoutCurrencyForm.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    /**
+     * Sets the selection and the amount accordingly to the value passed to the input
+     * @param newAmount - Changed amount from user input
+     */
+    const setNewAmount = useCallback(
+        (newAmount: string) => {
+            // Remove spaces from the newAmount value because Safari on iOS adds spaces when pasting a copied value
+            // More info: https://github.com/Expensify/App/issues/16974
+            const newAmountWithoutSpaces = stripSpacesFromAmount(newAmount);
+            const replacedCommasAmount = replaceCommaWithDot(newAmountWithoutSpaces);
```
Comment (luacmartins) [inline src/components/AmountWithoutCurrencyForm.tsx:31]:
```suggestion
            const replacedCommasAmount = replaceCommasWithPeriod(newAmountWithoutSpaces);
```

File: src/components/AmountWithoutCurrencyForm.tsx [hunk 2]
```diff
@@ -0,0 +1,61 @@
+import React, {useCallback, useMemo} from 'react';
+import useLocalize from '@hooks/useLocalize';
+import {replaceAllDigits, stripSpacesFromAmount, validateAmount} from '@libs/MoneyRequestUtils';
+import {replaceCommaWithDot} from '@libs/SearchUtils';
```
Comment (luacmartins) [inline src/components/AmountWithoutCurrencyForm.tsx:4]:
```suggestion
```

File: src/components/AmountWithoutCurrencyForm.tsx [hunk 3]
```diff
@@ -0,0 +1,61 @@
+import React, {useCallback, useMemo} from 'react';
+import useLocalize from '@hooks/useLocalize';
+import {replaceAllDigits, stripSpacesFromAmount, validateAmount} from '@libs/MoneyRequestUtils';
```
Comment (luacmartins) [inline src/components/AmountWithoutCurrencyForm.tsx:3]:
```suggestion
import {replaceAllDigits, replaceCommasWithPeriod, stripSpacesFromAmount, validateAmount} from '@libs/MoneyRequestUtils';
```

File: src/libs/SearchUtils.ts [hunk 4]
```diff
@@ -74,6 +74,10 @@ function getTransactionItemCommonFormattedProperties(
     };
 }
 
+function replaceCommaWithDot(amount: string): string {
+    return amount.replace(/,/g, '.');
+}
+
```
Comment (luacmartins) [inline src/libs/SearchUtils.ts:80]: Let's reuse [MoneyRequestUtils.replaceCmmasWithPeriod](https://github.com/Expensify/App/blob/4c3aceb78809e751ed58fcdac2630f162e04a3dd/src/libs/MoneyRequestUtils.ts#L20)
```suggestion
```

File: src/libs/SearchUtils.ts [hunk 5]
```diff
@@ -574,4 +601,5 @@ export {
     normalizeQuery,
     shouldShowYear,
     getExpenseTypeTranslationKey,
+    replaceCommaWithDot,
```
Comment (luacmartins) [inline src/libs/SearchUtils.ts:604]:
```suggestion
```

File: src/libs/SearchUtils.ts [hunk 6]
```diff
@@ -427,6 +431,28 @@ function buildDateFilterQuery(filterValues: Partial<SearchAdvancedFiltersForm>)
...
+    let amountFilter = '';
+    if (lessThan) {
+        amountFilter += `${CONST.SEARCH.SYNTAX_FILTER_KEYS.AMOUNT}<${lessThan}`;
+    }
+    if (lessThan && greaterThan) {
+        amountFilter += ' ';
+    }
+    if (greaterThan) {
+        amountFilter += `${CONST.SEARCH.SYNTAX_FILTER_KEYS.AMOUNT}>${greaterThan}`;
+    }
```
Comment (luacmartins) [inline src/libs/SearchUtils.ts:451]: Nit pick, let's invert the order so we show it always as `greaterThan` `lessThan`
```suggestion
    let amountFilter = '';
    if (greaterThan) {
        amountFilter += `${CONST.SEARCH.SYNTAX_FILTER_KEYS.AMOUNT}>${greaterThan}`;
    }
    if (lessThan && greaterThan) {
        amountFilter += ' ';
    }
    if (lessThan) {
        amountFilter += `${CONST.SEARCH.SYNTAX_FILTER_KEYS.AMOUNT}<${lessThan}`;
    }
```

**Review by luacmartins — CHANGES_REQUESTED** [review]
Looking good. Left a few comments.

**Comment by dubielzyk-expensify** [comment]
This is looking good to me. Could we see an example of where we have just `Greater than` filled in and one with just `Less than` filled in?

**Comment by luacmartins** [comment]
@rayane-djouah is that bug specific to this filter or does it happen in other text inputs?

**Review by luacmartins — DISMISSED** [review]
LGTM

**Comment by luacmartins** [comment]
@SzymczakJ could you please take a look at https://github.com/Expensify/App/pull/47343#issuecomment-2291467068?

**Comment by 289Adam289** [comment]
@SzymczakJ is OOO this week. I am currently working on a fix.

**Comment by dubielzyk-expensify** [comment]
Looking great from a visual perspective too

**Comment by 289Adam289** [comment]
@luacmartins This bug appears to be bigger than I originally thought. It is connected to these issues: https://github.com/facebook/react-native/issues/38155, https://github.com/facebook/react-native/issues/24585, and how the native TextInput component works. A similar problem has already been encountered in the Expensify App https://github.com/Expensify/App/issues/41961. I propose we deal with this in a follow-up issue.

**Comment by luacmartins** [comment]
That sounds good to me. @rayane-djouah wanna do a final review? We'll address the input issue in a follow up

**Comment by dylanexpensify** [comment]
bump @rayane-djouah for final review 🙇‍♂️

**Review by rayane-d — CHANGES_REQUESTED** [review]
Regarding the issue discussed here: https://github.com/Expensify/App/pull/47343#issuecomment-2288270209, I believe we need to multiply the amounts by 100 when building the search query to ensure they are correctly handled by the backend. However, The amounts will be correctly stored in Onyx with decimals and will be displayed correctly on the amount filter page and menu. We can determine how these amounts should be displayed in the search results page header as part of https://github.com/Expensify/App/issues/47254.

**Review by rayane-d — CHANGES_REQUESTED** [review]
I believe it’s better to store the amounts in Onyx in the backend format using `convertToBackendAmount`. For displaying the amounts in input default values and menu option preview text, we should use `convertToDisplayStringWithoutCurrency` to ensure the amounts are formatted and localized correctly.

**Review by rayane-d — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



https://github.com/user-attachments/assets/6fb6f13f-6b6d-43a3-aaaa-dea926c319e2


</details>

<details>
<summary>Android: mWeb Chrome</summary>



https://github.com/user-attachments/assets/541ded1e-c7fd-41c0-b205-e5006e6a6c63


</details>

<details>
<summary>iOS: Native</summary>



https://github.com/user-attachments/assets/e306d26b-71e3-453f-816b-82da9c486662


</details>

<details>
<summary>iOS: mWeb Safari</summary>



https://github.com/user-attachments/assets/b0d2c7db-a27d-4e74-8e6c-17ca1851bd45


</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



https://github.com/user-attachments/assets/25f2dbb3-a8b2-4586-bd90-9d8e70de4ba9


</details>

<details>
<summary>MacOS: Desktop</summary>



https://github.com/user-attachments/assets/392bd095-6d5f-4c2f-a3f3-2934953b7c30


</details>

**Review by luacmartins — APPROVED** [review]
LGTM

**Pull request merged**